### PR TITLE
The codes of Cloud-Robotic AI Benchmarking for Edge-cloud Collaborative Lifelong Learning

### PR DESCRIPTION
**Pull request by hsj576**
**What type of PR is this?**
This PR provides codes for PR#62

**What this PR does / why we need it**:
It provides a benchmark example of semantic segmentation for lifelong learning in Ianvs.

### CODE REVIEW

**Comment by MooreZheng** [comment]
@luosiqi might want to take a look at it

File: examples/curb-detection/lifelong_learning_bench/README.md [hunk 1]
```diff
@@ -26,7 +26,7 @@ cd /ianvs #One might use another path preferred
 
 mkdir project
 cd project
-git clone https://github.com/kubeedge/ianvs.git   
+git clone -b feature-lifelong-n https://github.com/kubeedge/ianvs.git
```
Comment (MooreZheng) [inline examples/curb-detection/lifelong_learning_bench/README.md:29]: In the formal version, the clone shall revised to the main branch

Comment (hsj576) [reply examples/curb-detection/lifelong_learning_bench/README.md:29]: Ok, I will revise it.

File: core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py [hunk 2]
```diff
@@ -82,7 +82,9 @@ def run(self):
         samples_transfer_ratio_info = self.system_metric_info.get(
             SystemMetricType.SAMPLES_TRANSFER_RATIO.value)
         mode = self.model_eval_config.get("model_metric").get("mode")
-        if mode != 'multi-inference':
+
+        # in this mode, the inference period is skipped to accelerate training speed
+        if mode == 'no-inference':
```
Comment (luosiqi) [inline core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py:87]: Is this code for test not for formal running? If it is, please modify it.

File: core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py [hunk 3]
```diff
@@ -111,19 +113,72 @@ def run(self):
...
                                                               "test")
 
-        samples_transfer_ratio_info.append((test_res, unseen_task_train_samples.x))
+            samples_transfer_ratio_info.append((test_res, unseen_task_train_samples.x))
+            '''
+            job = self.build_paradigm_job(ParadigmType.LIFELONG_LEARNING.value)
+            inference_dataset = self.dataset.load_data(self.dataset.test_url, "eval",
+                                                   feature_process=_data_feature_process)
+            kwargs = {}
+            #model_eval_info = self.model_eval_config
```
Comment (luosiqi) [inline core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py:134]: Annotations should be removed.

File: core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py [hunk 4]
```diff
@@ -146,18 +201,23 @@ def _inference(self, edge_task_index, data_index_file, rounds):
...
         unseen_task_labels = []
         mode = self.model_eval_config.get("model_metric").get("mode")
-        kwargs = {"mode": mode}
+        if mode is None:
+            kwargs = {} # fix the bug of "TypeError: call() got an unexpected keyword argument 'mode'"
+        else:
+            kwargs = {"mode": mode}
+        #print(len(inference_dataset.x))
         for i, _ in enumerate(inference_dataset.x):
+            #print(i)
```
Comment (luosiqi) [inline core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py:210]: Annotations should be deleted.

File: core/testenvmanager/dataset/dataset.py [hunk 5]
```diff
@@ -79,10 +79,35 @@ def _process_txt_index_file(cls, file_url):
             tmp_file = os.path.join(tempfile.mkdtemp(), "index.txt")
             with open(tmp_file, "w", encoding="utf-8") as file:
                 for line in lines:
-                    front, back = line.split(" ")
-                    file.writelines(
-                        f"{os.path.abspath(os.path.join(root, front))} "
-                        f"{os.path.abspath(os.path.join(root, back))}")
+                    #front, back = line.split(" ")
```
Comment (luosiqi) [inline core/testenvmanager/dataset/dataset.py:82]: Annotations should be removed.

File: core/testenvmanager/dataset/dataset.py [hunk 6]
```diff
@@ -148,6 +173,12 @@ def split_dataset(self, dataset_url, dataset_format, ratio, method="default",
                                               data_types=dataset_types,
                                               output_dir=output_dir,
                                               times=times)
+        # add new splitting method for semantic segmantation
+        if method == "my_splitting":
+            return self._my_splitting(dataset_url, dataset_format, ratio,
```
Comment (luosiqi) [inline core/testenvmanager/dataset/dataset.py:178]: The function name "_my_splitting" is not explicit.

File: examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/dataloaders/__init__.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            val_loader, num_class = None, cityscapes.CityscapesSegmentation.NUM_CLASSES
+
+        if test_data is not None:
+            test_set = cityscapes.CityscapesSegmentation(args, data=test_data, split='test')
+            num_class = test_set.NUM_CLASSES
+            test_loader = DataLoader(test_set, batch_size=args.test_batch_size, shuffle=False, **kwargs)
+        else:
+            test_loader, num_class = None, cityscapes.CityscapesSegmentation.NUM_CLASSES
+
+        # custom_set = cityscapes.CityscapesSegmentation(args, split='custom_resize')
```
Comment (luosiqi) [inline examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/dataloaders/__init__.py:28]: Annotations should be removed.

File: examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/dataloaders/datasets/cityscapes.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+        print("Found %d %s RGB images" % (len(self.images[split]), split))
+        print("Found %d %s disparity images" % (len(self.disparities[split]), split))
+
+
+    def __len__(self):
+        return len(self.images[self.split])
+
+    def __getitem__(self, index):
+        #print(type(self.images[self.split][index]))
```
Comment (luosiqi) [inline examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/dataloaders/datasets/cityscapes.py:59]: Annotations should be removed.

File: examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/eval.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    for name, param in state_dict.items():
+        if name not in own_state:
+            print('{} not in model_state'.format(name))
+            continue
+        else:
+            own_state[name].copy_(param)
+
+    return model
+
+# def main():
```
Comment (luosiqi) [inline examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/eval.py:197]: Annotations should be removed.

**Comment by luosiqi** [comment]
I suggest that the commits can be compressed by using "git rebase -i".

**Comment by luosiqi** [comment]
As you have modified core codes of ianvs, make sure it works for other paradigms such as single task learning and incremental learning.

File: core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py (hunk 2 above)
Comment (hsj576) [reply core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py:87]: This code provide multiple modes for running, users could choose either test or formal running in "algorithm.yaml".

File: core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py (hunk 3 above)
Comment (hsj576) [reply core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py:134]: Ok, I will remove annotations.

File: core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py (hunk 4 above)
Comment (hsj576) [reply core/testcasecontroller/algorithm/paradigm/lifelong_learning/lifelong_learning.py:210]: Ok, I will remove annotations.

File: core/testenvmanager/dataset/dataset.py (hunk 5 above)
Comment (hsj576) [reply core/testenvmanager/dataset/dataset.py:82]: Ok, I will remove annotations.

File: examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/dataloaders/__init__.py (hunk 7 above)
Comment (hsj576) [reply examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/dataloaders/__init__.py:28]: Ok, I will remove annotations.

File: examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/dataloaders/datasets/cityscapes.py (hunk 8 above)
Comment (hsj576) [reply examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/dataloaders/datasets/cityscapes.py:59]: Ok, I will remove annotations.

File: examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/eval.py (hunk 9 above)
Comment (hsj576) [reply examples/semantic_segmentation/lifelong_learning_bench/testalgorithms/rfnet/RFNet/eval.py:197]: Ok, I will remove annotations.

**Comment by hsj576** [comment]
> I suggest that the commits can be compressed by using "git rebase -i".

OK, I have compressed the commits.

File: core/testenvmanager/dataset/dataset.py (hunk 6 above)
Comment (hsj576) [reply core/testenvmanager/dataset/dataset.py:178]: Ok, I will change the name of the function

**Comment by MooreZheng** [comment]
It seems that the CI testing is not passed.

**Comment by luosiqi** [comment]
Please modify your codes according to CI. It seems that your codes don't meet the coding standard.

**Comment by hsj576** [comment]
> Please modify your codes according to CI. It seems that your codes don't meet the coding standard.

I have modified the codes to meet the coding standard.

**Pull request merged**